### PR TITLE
fixing sub-property externals case for validation only

### LINKED ISSUE #5631: Feature Request: externals should be able to specify arbitrary expressions
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Currently, externals only allow specifying a single variable as the value to use from the global scope. Eg `{externals: {'gem/Style':'Gem.Style'}` will return a bundle that loads `root["Gem.Style"]`. This is obviously not what I expected to load here.

**What is the expected behavior?**

I'd like to be able to specify an arbitrary expression in order to find the module. For example, I'd like to be able to specify externals like  `{externals: {"gem/Style": {expression: ".Gem.Style"}}}` so that webpack would render the external as `root.Gem.Style`. This would also allow arbitrary module loading in whatever form the user needs.

**If this is a feature request, what is motivation or use case for changing the behavior?**

My motivation is that I have a module that expects modules to be loaded individually in common js, like this:

```
var Text =require("gem/Text")
var CheckBox = require("gem/CheckBox")
```

but when loaded as a UMD module, expects the user to access those same things via:

```
var Text = Gem.Text
var CheckBox = Gem.CheckBox
```

Because I'm loading modules that don't exist in a single variable, I'm currently forced to work around this by manually creating some boiler plate that translates between these things, by having a configuration like: `{externals: {"gem/Style":"Style"}}` and adding `var Style = Gem.Style` at the top of my html page. 

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

Webpack v3.5.5

### PR DESCRIPTION
**Pull request by fresheneesz**
(eg {externals:{external:['a','b']}} ). Note that the actual handling of sub-properties still isn't working.

**What kind of change does this PR introduce?**

Bugfix

**Did you add tests for your changes?**

Yes

**If relevant, link to documentation update:**

N/A

**Summary**

fixes #5631

**Does this PR introduce a breaking change?**

No

**Other information**

This shouldn't be merged in without properly implementing externals sub-properties.

### CODE REVIEW

File: schemas/webpackOptionsSchema.json [hunk 1]
```diff
@@ -823,7 +823,8 @@
       "$ref": "#/definitions/entry"
     },
     "externals": {
-      "$ref": "#/definitions/externals"
+      "type": "object",
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]:
``` js
externals: /xyz/
externals: ["a", "b"]
externals: "xyz"
```

These are also valid options.

Comment (fresheneesz) [reply schemas/webpackOptionsSchema.json]: Ah ok. I'll correct that

Comment (fresheneesz) [reply schemas/webpackOptionsSchema.json]: Actually, what does the second two mean? I don't see those forms documented or tested:

```
externals: ["a", "b"]
externals: "xyz"
```

Especially with the array, its ambiguous to me whether its defining that both "a" and "b" are external modules, or whether it means a sub property corresponding to `root["a"]["b"]`.

Comment (fresheneesz) [reply schemas/webpackOptionsSchema.json]: Ok no I got it. `["a","b"]` is defining both of those as external modules. Will fix it.

**Comment by fresheneesz** [comment]
Ok updated the schema to support the following forms:
```
externals: /x/
externals: "x"
externals: function(){}
externals: ["a","b",/c/, function(){}, {}]
externals: {
 a: "a",
 "b/c": ["b","c"],
 d: {
   umd: "d",
   root: ["_","d"]
 }
}
```

**Comment by fresheneesz** [comment]
@webpack-bot Alright then, didn't realize that mattered.

File: schemas/webpackOptionsSchema.json [hunk 2]
```diff
@@ -75,6 +75,20 @@
...
+        {
+          "type": "boolean"
+        },
+        {
+          "type": "array"
+        }
+      ],
+      "description": "Represents a replacement object for an external modules (configured by `externals`)"
+    },
     "externals": {
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]: external**s** is now unused?

File: schemas/webpackOptionsSchema.json [hunk 3]
```diff
@@ -108,7 +112,17 @@
         },
         {
           "items": {
-            "$ref": "#/definitions/externals"
+            "anyOf": [
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]: now

``` js
externals: [
  /abc/,
  () => {}
]
```

is no longer allowed?

File: schemas/webpackOptionsSchema.json [hunk 4]
```diff
@@ -83,17 +97,7 @@
...
-                "type": "string"
-              },
-              {
-                "type": "object"
-              },
-              {
-                "type": "boolean"
-              }
-            ]
+            "$ref": "#/definitions/external"
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]:
``` js
externals: {
  abc: {
    root: "Abc",
    commonjs: "abc"
  }
}
```

is no longer allowed?

File: schemas/webpackOptionsSchema.json (hunk 2 above)
Comment (fresheneesz) [reply schemas/webpackOptionsSchema.json]: Oops, that wasn't intentional

**Comment by fresheneesz** [comment]
@sokra Ok I added some additional test cases and fixed the webpackOptionsSchema so that it doesn't seem to be getting validation errors. I'm surprised, tho, that the tests seem to be failing *after* validation. Maybe I wrote the tests wrong?

**Comment by vankop** [comment]
Thanks for your contribution. I am closing PR because it is too stale..

**Comment by fresheneesz** [comment]
Sorry, it seems I lost track of this. Is this PR still useful? Should I try to fix the issues?

**Pull request closed without merging**